### PR TITLE
Jetty 12.0.x deployment

### PR DESCRIPTION
**Pull request by gregw**
Deploy webapps for different environments from the same webapps directory.
The maximal environment known to the AppProviders is used as the default environment.
An explicit environment can be set in a properties file for an application, which is also used for property substitution in any context xml file.

### CODE REVIEW

File: jetty-core/jetty-deploy/pom.xml [hunk 1]
```diff
@@ -56,5 +56,13 @@
       <artifactId>jetty-test-helper</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (janbartel) [inline jetty-core/jetty-deploy/pom.xml:59]: aren't these identical?

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java [hunk 2]
```diff
@@ -13,36 +13,60 @@
...
+
+import org.eclipse.jetty.deploy.util.FileID;
 import org.eclipse.jetty.server.handler.ContextHandler;
-import org.eclipse.jetty.util.Attributes;
 
 /**
- * The information about an App that is managed by the {@link DeploymentManager}
+ * The information about an App that is managed by the {@link DeploymentManager}.
+ *
+ * </p>
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java]: Unneeded para

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java [hunk 3]
```diff
@@ -117,9 +134,16 @@ public String getContextPath()
...
+    /**
+     * Get the environment name.
+     * If the property "environment" exists, then that is returned as the environment, otherwise
+     * the {@link DeploymentManager#getDefaultEnvironmentName()} is returned.
+     * @return The {@link org.eclipse.jetty.util.component.Environment} name for the application.
+     */
+    public String getEnvironmentName()
     {
-        return _environment;
+        String name = getProperties().get("environmnet");
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java]: Misspelling of environment

**Review by janbartel — CHANGES_REQUESTED** [review]

**Review by janbartel — CHANGES_REQUESTED** [review]
Something is wrong, not sure what. I checked out your branch, did a local build, then made a base and tried to deploy `ee10-demo`, which failed.  Here's the transcript.

janb@graphene: ~/src/jetty-eclipse/jetty-12.0.x/jetty-home/target/base (jetty-12.0.x-deployment)
[2154] java -jar ../start.jar --add-module=ee10-demo
Error: Unable to access jarfile ../start.jar

janb@graphene: ~/src/jetty-eclipse/jetty-12.0.x/jetty-home/target/base (jetty-12.0.x-deployment)
[2155] java -jar ../jetty-home/start.jar --add-module=ee10-demo
INFO  : mkdir ${jetty.base}/start.d
INFO  : server          transitively enabled, ini template available with --add-module=server
INFO  : alpn-java       transitively enabled
INFO  : alpn            transitively enabled, ini template available with --add-module=alpn
INFO  : test-keystore   transitively enabled
INFO  : ee10-demo-rewrite transitively enabled
INFO  : ee10-demo       initialized in ${jetty.base}/start.d/ee10-demo.ini
INFO  : ee10-demo-jndi  transitively enabled
INFO  : jdbc            transitively enabled
INFO  : ee10-jaas       transitively enabled, ini template available with --add-module=ee10-jaas
INFO  : threadpool      transitively enabled, ini template available with --add-module=threadpool
INFO  : ssl             transitively enabled, ini template available with --add-module=ssl
INFO  : ee10-servlets   transitively enabled
INFO  : ee10-demo-root  transitively enabled
INFO  : deploy          transitively enabled
INFO  : logging-jetty   transitively enabled
INFO  : ee10-deploy     transitively enabled, ini template available with --add-module=ee10-deploy
INFO  : ee10-webapp     transitively enabled, ini template available with --add-module=ee10-webapp
INFO  : ee10-plus       transitively enabled
INFO  : ee10-demo-jetty transitively enabled
INFO  : client          transitively enabled
INFO  : ee10-demo-mock-resources transitively enabled
INFO  : http2           transitively enabled, ini template available with --add-module=http2
INFO  : ee10-glassfish-jstl transitively enabled
INFO  : https           transitively enabled
INFO  : bytebufferpool  transitively enabled, ini template available with --add-module=bytebufferpool
INFO  : ee10-websocket-jetty transitively enabled
INFO  : ee10-annotations transitively enabled
INFO  : ext             transitively enabled
INFO  : sessions        transitively enabled, ini template available with --add-module=sessions
INFO  : ee10-demo-async-rest transitively enabled
INFO  : ee10-security   transitively enabled
INFO  : work            transitively enabled
INFO  : ee10-demo-moved-context transitively enabled
INFO  : ee10-demo-proxy transitively enabled
INFO  : ee10-demo-spec  transitively enabled
INFO  : resources       transitively enabled
INFO  : ee10-demo-jaas  transitively enabled
INFO  : rewrite         transitively enabled, ini template available with --add-module=rewrite
INFO  : ee10-apache-jsp transitively enabled
INFO  : ee10-jstl       transitively enabled
INFO  : ee10-demo-jsp   transitively enabled
INFO  : jndi            transitively enabled
INFO  : ee10-websocket-jakarta transitively enabled
INFO  : ee10-demo-realm transitively enabled, ini template available with --add-module=ee10-demo-realm
INFO  : http            transitively enabled, ini template available with --add-module=http
INFO  : ee10-servlet    transitively enabled
INFO  : ee10-jsp        transitively enabled
INFO  : logging/slf4j   dynamic dependency of logging-jetty
INFO  : mkdir ${jetty.base}/lib/bouncycastle
INFO  : copy /home/janb/.m2/repository/org/bouncycastle/bcpkix-jdk15on/1.70/bcpkix-jdk15on-1.70.jar to ${jetty.base}/lib/bouncycastle/bcpkix-jdk15on-1.70.jar
INFO  : copy /home/janb/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.70/bcprov-jdk15on-1.70.jar to ${jetty.base}/lib/bouncycastle/bcprov-jdk15on-1.70.jar
INFO  : download https://repo1.maven.org/maven2/org/bouncycastle/bcutil-jdk15to18/1.70/bcutil-jdk15to18-1.70.jar to ${jetty.base}/lib/bouncycastle/bcutil-jdk15on-1.70.jar
INFO  : mkdir ${jetty.base}/etc
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-rewrite-rules.xml to ${jetty.base}/etc/ee10-demo-rewrite-rules.xml
INFO  : mkdir ${jetty.base}/webapps-ee10
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-jndi.xml to ${jetty.base}/webapps-ee10/ee10-demo-jndi.xml
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jndi-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jndi-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jndi-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jndi-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : mkdir ${jetty.base}/lib/ext
INFO  : copy /home/janb/.m2/repository/jakarta/mail/jakarta.mail-api/2.0.0/jakarta.mail-api-2.0.0.jar to ${jetty.base}/lib/ext/jakarta.mail-api-2.0.0.jar
INFO  : mkdir ${jetty.base}/webapps-ee10/root-ee10
INFO  : mkdir ${jetty.base}/webapps-ee10/root-ee10/images
INFO  : copy ${jetty.home}/modules/demo.d/ee10-root/index.html to ${jetty.base}/webapps-ee10/root-ee10/index.html
INFO  : copy ${jetty.home}/modules/demo.d/ee10-root/jetty.css to ${jetty.base}/webapps-ee10/root-ee10/jetty.css
INFO  : copy ${jetty.home}/modules/demo.d/ee10-root/images/jetty-pic.png to ${jetty.base}/webapps-ee10/root-ee10/images/jetty-pic.png
INFO  : copy ${jetty.home}/modules/demo.d/ee10-root/images/webtide_logo.jpg to ${jetty.base}/webapps-ee10/root-ee10/images/webtide_logo.jpg
INFO  : mkdir ${jetty.base}/resources
INFO  : copy ${jetty.home}/modules/logging/jetty/resources/jetty-logging.properties to ${jetty.base}/resources/jetty-logging.properties
INFO  : mkdir ${jetty.base}/webapps
INFO  : mkdir ${jetty.base}/webapps-ee10/ee10-demo-jetty.d
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-jetty.xml to ${jetty.base}/webapps-ee10/ee10-demo-jetty.xml
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-jetty-override-web.xml to ${jetty.base}/webapps-ee10/ee10-demo-jetty.d/ee10-demo-jetty-override-web.xml
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jetty-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jetty-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jetty-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jetty-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-mock-resources/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-mock-resources/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-mock-resources/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-mock-resources/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-async-rest-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-async-rest-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-async-rest-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-async-rest-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : mkdir ${jetty.base}/work
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-moved-context.xml to ${jetty.base}/webapps-ee10/ee10-demo-moved-context.xml
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-proxy-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-proxy-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-proxy-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-proxy-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-spec.xml to ${jetty.base}/webapps-ee10/ee10-demo-spec.xml
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-spec-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-spec-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-spec-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-spec-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-jaas.xml to ${jetty.base}/webapps-ee10/ee10-demo-jaas.xml
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-login.conf to ${jetty.base}/etc/ee10-demo-login.conf
INFO  : copy ${jetty.home}/modules/demo.d/ee10-demo-login.properties to ${jetty.base}/etc/ee10-demo-login.properties
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jaas-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jaas-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jaas-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jaas-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : copy ${jetty.home}/modules/rewrite/jetty-rewrite-rules.xml to ${jetty.base}/etc/jetty-rewrite-rules.xml
INFO  : mkdir /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jsp-webapp/12.0.0-SNAPSHOT
INFO  : download https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jsp-webapp/12.0.0-SNAPSHOT/maven-metadata.xml to /home/janb/.m2/repository/org/eclipse/jetty/ee10/demos/ee10-demo-jsp-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
java.io.IOException: URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jsp-webapp/12.0.0-SNAPSHOT/maven-metadata.xml
	at org.eclipse.jetty.start.FileInitializer.download(FileInitializer.java:142)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.download(MavenLocalRepoFileInitializer.java:306)
	at org.eclipse.jetty.start.fileinits.MavenLocalRepoFileInitializer.create(MavenLocalRepoFileInitializer.java:207)
	at org.eclipse.jetty.start.BaseBuilder.processFileResource(BaseBuilder.java:371)
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:408)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)
INFO  : copy ${jetty.home}/modules/demo.d/jetty-ee10-demo-realm.xml to ${jetty.base}/etc/jetty-ee10-demo-realm.xml
INFO  : copy ${jetty.home}/modules/demo.d/jetty-ee10-demo-realm.properties to ${jetty.base}/etc/jetty-ee10-demo-realm.properties
WARN  : Failed to process all file resources.
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jndi-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jndi.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jetty-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jetty.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-mock-resources/12.0.0-SNAPSHOT/maven-metadata.xml - lib/ee10-demo-mock-resources-12.0.0-SNAPSHOT.jar
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-async-rest-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-async-rest.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-proxy-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-proxy.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-spec-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-spec.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jaas-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jaas.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jsp-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jsp.war
java.lang.RuntimeException: Failed to process all file resources.
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jndi-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jndi.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jetty-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jetty.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-mock-resources/12.0.0-SNAPSHOT/maven-metadata.xml - lib/ee10-demo-mock-resources-12.0.0-SNAPSHOT.jar
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-async-rest-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-async-rest.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-proxy-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-proxy.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-spec-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-spec.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jaas-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jaas.war
 - [IOException] URL GET Failure [404/Not Found] on https://repo1.maven.org/maven2/org/eclipse/jetty/ee10/demos/ee10-demo-jsp-webapp/12.0.0-SNAPSHOT/maven-metadata.xml - webapps-ee10/ee10-demo-jsp.war
	at org.eclipse.jetty.start.BaseBuilder.processFileResources(BaseBuilder.java:428)
	at org.eclipse.jetty.start.BaseBuilder.build(BaseBuilder.java:339)
	at org.eclipse.jetty.start.Main.start(Main.java:453)
	at org.eclipse.jetty.start.Main.main(Main.java:75)

Usage: java -jar $JETTY_HOME/start.jar [options] [properties] [configs]
       java -jar $JETTY_HOME/start.jar --help  # for more information

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java [hunk 4]
```diff
@@ -117,9 +132,16 @@ public String getContextPath()
...
+    /**
+     * Get the environment name.
+     * If the property "environment" exists, then that is returned as the environment, otherwise
+     * the {@link DeploymentManager#getDefaultEnvironmentName()} is returned.
+     * @return The {@link org.eclipse.jetty.util.component.Environment} name for the application.
+     */
+    public String getEnvironmentName()
     {
-        return _environment;
+        String name = getProperties().get("environment");
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java]: Maybe make a constant for that property name.

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 5]
```diff
@@ -55,27 +51,31 @@
...
+ * </p>
  * <p>For XML configured contexts, the ID map will contain a reference to the {@link Server} instance called "Server" and
- * properties for the webapp file as "jetty.webapp" and directory as "jetty.webapps".
+ * properties for the webapp file as "jetty.webapp" and directory as "jetty.webapps".  The properties will be initialized
+ * with:<ul>
+ *     <li>The properties set on the application via {@link App#getProperties()}; otherwise:</li>
+ *     <li>The properties set on this provider via {@link #getProperties()}</li>
+ * </ul>
  */
 @ManagedObject("Provider for start-up deployement of webapps based on presence in directory")
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]:
```suggestion
@ManagedObject("Provider for start-up deployment of webapps based on presence in directory")
```

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 6]
```diff
@@ -131,161 +136,186 @@ public ContextProvider()
...
      */
     public void setDefaultsDescriptor(String defaultsDescriptor)
     {
-        _defaultsDescriptor = defaultsDescriptor;
-    }
-
-    public ConfigurationManager getConfigurationManager()
-    {
-        return _configurationManager;
+        _properties.put("jetty.deploy.defaultsDescriptor", defaultsDescriptor);
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: If you're going to be referencing the same property a couple of times in this file, then make a constant for it. In general referencing property names from xml files is a little fragile, but I understand why it's being done. I guess ide refactoring will make any future changes to those property names less problematic.

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 7]
```diff
@@ -131,161 +136,186 @@ public ContextProvider()
...
+    /**
+     *
+     * This is equivalent to the <code>jetty.deploy.configurationClasses</code> property.
+     * @return The configuration class names.
+     */
     @ManagedAttribute("configuration classes for webapps to be processed through")
     public String[] getConfigurationClasses()
     {
-        return _configurationClasses;
+        String cc = _properties.get("jetty.deploy.configurationClasses");
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: Same comment: make a constant.

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 8]
```diff
@@ -131,161 +136,186 @@ public ContextProvider()
...
-    public void setConfigurationManager(ConfigurationManager configurationManager)
+    public void setConfigurationClasses(String configurations)
     {
-        updateBean(_configurationManager, configurationManager);
-        _configurationManager = configurationManager;
+        setConfigurationClasses(StringUtil.isBlank(configurations) ? null : configurations.split(","));
     }
 
     /**
+     * This is equivalent to the <code>jetty.deploy.configurationClasses</code> property.
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: Confusing? Is this overriding the `jetty.deploy.configurationClasses` property or setting it?

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 9]
```diff
@@ -131,161 +136,186 @@ public ContextProvider()
...
+     * <p>
+     * This is equivalent to the <code>jetty.deploy.tempDir</code> property.
+     * If not set, then the <code>java.io.tmpdir</code> System Property is used.
      *
      * @param directory the new work directory
      */
     public void setTempDir(File directory)
     {
-        _tempDirectory = directory;
+        _properties.put("jetty.deploy.tempDir", directory.getAbsolutePath());
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java]: Constant?

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java (hunk 4 above)
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java]: Is this something that should be known in the constructor and immutable thereafter?

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java [hunk 10]
```diff
@@ -61,6 +83,11 @@ public AppProvider getAppProvider()
         return _provider;
     }
 
+    public Map<String, String> getProperties()
```
Comment (janbartel) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java:87]: Should the properties be private to the `App`?

**Review by janbartel — APPROVED** [review]
Some questions, comments, opinions in the comments.

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java (hunk 10 above)
Comment (gregw) [reply jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java:87]: No, there is a use-case for a different provider mutating properties after construction.

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/App.java]: properties can be mutated after construction

**Pull request merged**